radar-auth/src/main/java/org/radarcns/auth/token/AbstractRadarToken.java

```diff
@@ -25,17 +27,23 @@ public boolean hasPermission(Permission permission) {
     @Override
     public boolean hasPermissionOnProject(Permission permission, String projectName) {
         return hasScope(permission.scopeName())
-                && (isClientCredentials() || hasAuthorityForPermission(permission, projectName));
+                && (isClientCredentials() || hasAuthorityForProject(permission, projectName));
     }
 
     @Override
     public boolean hasPermissionOnSubject(Permission permission, String projectName,
             String subjectName) {
         return hasScope(permission.scopeName())
-                && (isClientCredentials() || hasAuthorityForPermission(permission, projectName,
+                && (isClientCredentials() || hasAuthorityForSubject(permission, projectName,
                         subjectName));
     }
 
+    @Override
+    public boolean hasPermissionOnSource(Permission permission, String sourceId) {
```

Comment:
shouldn't we validate based on projectName too,e.g. similar to `hasPermissionOnSubject`?

Reply:
It's just that the project and source aren't directly linked. I don't know what that would validate...